### ISSUE TITLE
Conceptual issue with copying trainlib (project template vs. library)

### ISSUE THREAD

**Issue by markus-hinsche** [kbressem/trainlib#8]
Currently trainlib is used as a template for starting a project (like a cookiecutter template)

However, trainlib seems to become a library, s.t. for example transforms.py should just be copied to every project, but should not changed (so it is consistent with trainlib). 

Therefore, trainlib might be a libarary instead of a template. 
You can import it (`from trainlib import transforms`), no need to copy, and therefore DRY

(At this point in time, to make it into a library, trainlib might have to become a little more flexible, e.g. accept more data) or allow more configurations in config.yml files)

**Comment by kbressem** [comment]
I don't see it as a template but instead it's my training library, which aims to provide very quick results in segmentation projects. Instead of writing the same code over and over again in all projects, the aim with `trainlib` is to aggregate the code in one place and continually improve it. Copying any trainlib code to a new project is not the intended use, and not needed. `trainlib` can already be installed as a library with pip and will work out of the box for segmentation projects, given your data is in the expected format (thus closing this issue). That's how I use it for other projects.  
I agree that it must become a little bit more flexible (e.g., #5 is a problem). But I intend to move slow with this. Only if the problem appears in more than one project, I will incorporate it into `trainlib`, to keep the codebase as small as possible.

**Comment by markus-hinsche** [comment]
> trainlib can already be installed as a library with pip 

Can you document this in the REAMDE? Currently it says `pip install .`, I would expect something like `pip install trainlib` or `pip install https://github.com/kbressem/trainlib`

>  That's how I use it for other projects.

Can you link another project to the README which uses trainlib as a library?

> it's my training library
> But I intend to move slow with this. 

I think you have to consider if it is "your" library or if it is a group of contributors developing trainlib (I prefer the latter).

If it is "your" library and projects of different people (like me) use it, then I would like to change it or being able to develop without you around.

**Comment by kbressem** [comment]
> Can you document this in the REAMDE? Currently it says `pip install .`, I would expect something like `pip install trainlib` or `pip install https://github.com/kbressem/trainlib`

I will update it. 

> Can you link another project to the README which uses trainlib as a library?

Those are still private, but I will add you as a contributor. 

> I think you have to consider if it is "your" library or if it is a group of contributors developing trainlib (I prefer the latter).

With this I meant, It reflects how I want to approach things and projects and what I currently consider best practices. I started to work on `trainlib`, because I noticed, that I tend to repeat a lot of code between projects. For all segmentation projects, the same workflows, metrics, loss functions and data pipelines were implemented over and over again. Still, this usually took a lot of time, which I felt was essentially wasted as it prevented me from working on stuff that really mattered, such as model design and hyperparamter tuning. `trainlib` was designed to provide most of what is needed for these projects out of the box, so that I can start working on important/fun stuff faster.  
It was also made to keep my research reproducible. So I can come back to a project after 6 months and it sill works.  
Since writing `trainlib` I was able to move considerably faster and get first good segmentation models alongside metrics, extensive logs and visualizations within a few days after the start of a project.  
I am more than happy to share `trainlib` and work with others to improve the library, however, if this involves the few use-cases `trainlib` was intended for and uses the approaches I consider best practices. For example, this means I would not start to support JSON instead of CSV to store file names or add support for histopathology images. As collaborators on `trainlib` will most likely people who I collaborate on other projects with, this should work out.
